### PR TITLE
[NPM] NPM Debug CLI

### PR DESCRIPTION
**Pull request by nddq**
This PR contains the code for the NPM Debug CLI that interacts with the data plane parser + converter and traffic analyzer.

To use this CLI, simply build the NPM binaries and then execute it with the flag debug. i.e /usr/bin/azure-npm debug.

### CODE REVIEW

**Review by huntergregory — CHANGES_REQUESTED** [review]
Looks good Quang. Loved the demo you showed me earlier. Couple things:
1. You have some duplicate logic in your 3 cobra commands (parseIPtable, getTuples, convertIPtable) where you get the result and error from a function and then print those two out. Can you condense this? 
Example: 
```
if npmCacheF == "" && iptableSaveF == "" {
	ipTableRulesRes, err := c.GetJSONRulesFromIptables(iptableName)
	if err != nil {
		fmt.Printf("%+v\n", err)
	}
	fmt.Printf("%s\n", ipTableRulesRes)
} else {
	ipTableRulesRes, err := c.GetJSONRulesFromIptableFile(iptableName, npmCacheF, iptableSaveF)
	if err != nil {
		fmt.Printf("%+v\n", err)
	}
	fmt.Printf("%s\n", ipTableRulesRes)
}
```
Write this as
```
var result TYPE
var err error
if () { 
	result, err := c.Get(name)
} else {
	result, err := c.GetFromFile(name, cacheFile, saveF)
}
if err != nil {
	// print error
}
// print result
```
2. Do you have the team's go linter? You have some weird formatting like a new line before the end of a function sometimes, which GitHub doesn't like.
3. Could you skip the else bracket in plugin/main.go? Just say `return` after the if statement? Seems a little cleaner. 

Then just a couple questions for my own understanding:
1. When would there be more than os arg in main.go, and why do you skip the rest of the function when this is the case?
2. Why are we switching from klog to log? Also, did you mean to delete the "Start NPM version" log line?

**Comment by rbtr** [comment]
@huntergregory
> Write this as
> 
> ```
> var result TYPE
> var err error
> if () { 
> 	result, err := c.Get(name)
> } else {
> 	result, err := c.GetFromFile(name, cacheFile, saveF)
> }
> if err != nil {
> 	// print error
> }
> // print result
> ```

Whenever I see variable declarations moving further from their use, I get suspicious (that something more significant is wrong with the flow)... maybe once the branches of the if/else approach the size of the dedup or get over a handful of lines each, I would argue against this.

@nddq 
Since these cmds are in a `RunE`, you don't need to print errs here, you can just return them and let the root invocation log and exit. And I'm not sure of this, but I think all of `GetNetworkTupleFile` func args need to be valid, so should we `||` instead of `&&`?

**Comment by nddq** [comment]
> 1. Do you have the team's go linter? You have some weird formatting like a new line before the end of a function sometimes, which GitHub doesn't like.
> 2. Could you skip the else bracket in plugin/main.go? Just say `return` after the if statement? Seems a little cleaner.
> 
> Then just a couple questions for my own understanding:
> 
> 1. When would there be more than 1 os arg in main.go, and why do you skip the rest of the function when this is the case?
> 2. Why are we switching from klog to log? Also, did you mean to delete the "Start NPM version" log line?

1. Can you point out a case of weird formatting for me because I don't seem to find any. I run the golangci-lint every time I planned to commit the code so this should not be an issue.
2. Did you mean the else bracket after `if len(os.Args) > 1`? This is actually the entry point for NPM :) 
3. To loop back to my previous answer, I want to start the CLI only when there is a debug command right after calling the binaries, otherwise, it would start NPM itself. That's why I have a check for the os.Args.
4. This is my mistake, I some how had the old main.go file. I fixed that in my most recent commit!

**Comment by nddq** [comment]
@rbtr 
> Since these cmds are in a `RunE`, you don't need to print errs here, you can just return them and let the root invocation log and exit. And I'm not sure of this, but I think all of `GetNetworkTupleFile` func args need to be valid, so should we `||` instead of `&&`?

Yep you are correct, I will fix that.

File: npm/debugTools/cmd/debugCmd.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+var debugCmd = &cobra.Command{
+	Use:   "debug",
+	Short: "Collection of functions related to Azure NPM's debugging tools",
+}
+
+func init() {
+	rootCmd.AddCommand(debugCmd)
+	debugCmd.PersistentFlags().StringP("npmF", "n", "", "Set the NPM cache file path (optional)")
+	debugCmd.PersistentFlags().StringP("iptF", "i", "", "Set the iptable-save file path (optional)")
+	debugCmd.PersistentFlags().StringP("table", "t", "", "Set table name (default to filter)")
```
Comment (matmerr) [inline npm/debugTools/cmd/debugCmd.go:17]: If we add another cmd it might not require these flags, potentially reduce scope

**Review by matmerr — CHANGES_REQUESTED** [review]
Mostly ux changes 👍

File: npm/debugTools/cmd/root.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+// Execute adds all child commands to the root command and sets flags appropriately.
+// This is called by main.main(). It only needs to happen once to the rootCmd.
+func Execute() {
+	cobra.CheckErr(rootCmd.Execute())
+}
+
+func init() {
+	cobra.OnInitialize(initConfig)
+
+	rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", "config file (default is $HOME/.cli.yaml)")
```
[the remark this answers is not in the document]
Comment (rbtr) [reply npm/debugTools/cmd/root.go:28]: it is kind of a pain to set env vars on the fly if you want to test different configs though

[the remark this answers is not in the document]
Comment (matmerr) [reply npm/debugTools/cmd/root.go:28]: Perhaps, but some of the options that are being added here are on-disk targets, (npm cache file path, iptables-save file path) and if I need to test those on the fly I'll just set the npmF/iptF flag in the line for the cmd that needs it

The rest of options like src/dst/table are command specific and probably won't be in a config

**Review by matmerr — APPROVED** [review]
🚀

**Pull request merged**